src/dfx/src/lib/operations/canister/create_canister.rs

```diff
@@ -85,6 +87,71 @@ pub async fn create_canister(
         return reserve_canister_with_playground(env, canister_name).await;
     }
 
+    // Specified ID from the command line takes precedence over the one in dfx.json.
+    let mut specified_id = match (
+        config_interface.get_specified_id(canister_name)?,
+        specified_id_from_cli,
+    ) {
+        (Some(specified_id_from_json), Some(specified_id_from_cli)) => {
+            if specified_id_from_json != specified_id_from_cli {
+                warn!(
+                    env.get_logger(),
+                    "Canister '{0}' has a specified ID in dfx.json: {1},
+which is different from the one specified in the command line: {2}.
+The command line value will be used.",
+                    canister_name,
+                    specified_id_from_json,
+                    specified_id_from_cli
+                );
+            }
+            Some(specified_id_from_cli)
+        }
+        (Some(specified_id_from_json), None) => Some(specified_id_from_json),
+        (None, Some(specified_id_from_cli)) => Some(specified_id_from_cli),
+        (None, None) => None,
+    };
+
+    // If the network is IC mainnet, the specified ID will be overwritten by None.
+    if env.get_network_descriptor().is_ic && specified_id.is_some() {
+        warn!(
+            env.get_logger(),
+            "Specified ID is ignored on the IC mainnet."
+        );
+        specified_id = None;
+    }
+
+    // Replace call_sender with wallet canister unless:
```

Comment:
This comment describes the predicate below, and the flow, but I'm still not quite following.  Can you describe _why_ we need to replace the call sender with the wallet?